### PR TITLE
Switch to intra-doc links in core/src/{convert,iter}/mod.rs

### PR DESCRIPTION
**Pull request by nixphix**
Partial fix for #75080

@rustbot modify labels: T-doc, A-intra-doc-links, T-rustdoc

r? @jyn514

couldn't fix these

```rust
     ../../std/string/struct.String.html
     ../../std/primitive.never.html
     ../../std/sync/mpsc/struct.TryIter.html
```

### CODE REVIEW

File: library/core/src/convert/mod.rs [hunk 1]
```diff
@@ -141,13 +134,11 @@ pub const fn identity<T>(x: T) -> T {
...
 ///
-/// [`Option<T>`]: ../../std/option/enum.Option.html
-/// [`Result<T, E>`]: ../../std/result/enum.Result.html
-/// [`Borrow`]: ../../std/borrow/trait.Borrow.html
-/// [`Hash`]: ../../std/hash/trait.Hash.html
-/// [`Eq`]: ../../std/cmp/trait.Eq.html
-/// [`Ord`]: ../../std/cmp/trait.Ord.html
-/// [`&str`]: ../../std/primitive.str.html
+/// [`Option<T>`]: crate::option::Option
+/// [`Result<T, E>`]: crate::result::Result
```
Comment (jyn514) [inline library/core/src/convert/mod.rs:138]: These are in the prelude.
```suggestion
/// [`Option<T>`]: Option
/// [`Result<T, E>`]: Result
```

File: library/core/src/convert/mod.rs [hunk 2]
```diff
@@ -177,8 +168,8 @@ pub trait AsRef<T: ?Sized> {
 /// **Note: This trait must not fail**. If the conversion can fail, use a
 /// dedicated method which returns an [`Option<T>`] or a [`Result<T, E>`].
 ///
-/// [`Option<T>`]: ../../std/option/enum.Option.html
-/// [`Result<T, E>`]: ../../std/result/enum.Result.html
+/// [`Option<T>`]: crate::option::Option
+/// [`Result<T, E>`]: crate::result::Result
```
Comment (jyn514) [inline library/core/src/convert/mod.rs:172]:
```suggestion
/// [`Option<T>`]: Option
/// [`Result<T, E>`]: Result
```

File: library/core/src/convert/mod.rs [hunk 3]
```diff
@@ -204,7 +195,7 @@ pub trait AsRef<T: ?Sized> {
 /// assert_eq!(*boxed_num, 1);
 /// ```
 ///
-/// [`Box<T>`]: ../../std/boxed/struct.Box.html
+/// [`Box<T>`]: crate::boxed::Box<T>
```
Comment (jyn514) [inline library/core/src/convert/mod.rs:198]: This link will fail, `core` doesn't have access to `alloc`.

File: library/core/src/convert/mod.rs [hunk 4]
```diff
@@ -278,13 +269,10 @@ pub trait AsMut<T: ?Sized> {
 /// is_hello(s);
 /// ```
 ///
-/// [`TryInto`]: trait.TryInto.html
-/// [`Option<T>`]: ../../std/option/enum.Option.html
-/// [`Result<T, E>`]: ../../std/result/enum.Result.html
+/// [`Option<T>`]: crate::option::Option
+/// [`Result<T, E>`]: crate::result::Result
```
Comment (jyn514) [inline library/core/src/convert/mod.rs:273]:
```suggestion
/// [`Option<T>`]: Option
/// [`Result<T, E>`]: Result
```

File: library/core/src/convert/mod.rs [hunk 5]
```diff
@@ -278,13 +269,10 @@ pub trait AsMut<T: ?Sized> {
...
 /// [`String`]: ../../std/string/struct.String.html
-/// [`From`]: trait.From.html
-/// [`Into`]: trait.Into.html
-/// [`Vec`]: ../../std/vec/struct.Vec.html
+/// [`Vec`]: crate::vec::Vec<T>
```
Comment (jyn514) [inline library/core/src/convert/mod.rs:275]: This link will fail, `core` doesn't have access to `alloc`.

File: library/core/src/convert/mod.rs [hunk 6]
```diff
@@ -370,12 +358,9 @@ pub trait Into<T>: Sized {
 /// }
 /// ```
 ///
-/// [`TryFrom`]: trait.TryFrom.html
-/// [`Option<T>`]: ../../std/option/enum.Option.html
-/// [`Result<T, E>`]: ../../std/result/enum.Result.html
+/// [`Option<T>`]: crate::option::Option
+/// [`Result<T, E>`]: crate::result::Result
```
Comment (jyn514) [inline library/core/src/convert/mod.rs:362]:
```suggestion
/// [`Option<T>`]: Option
/// [`Result<T, E>`]: Result
```

File: library/core/src/convert/mod.rs [hunk 7]
```diff
@@ -370,12 +358,9 @@ pub trait Into<T>: Sized {
...
 /// [`String`]: ../../std/string/struct.String.html
-/// [`Into`]: trait.Into.html
-/// [`from`]: trait.From.html#tymethod.from
 /// [book]: ../../book/ch09-00-error-handling.html
```
Comment (jyn514) [inline library/core/src/convert/mod.rs:364]: `from` isn't a free function.
```suggestion
/// [`from`]: From::from
/// [book]: ../../book/ch09-00-error-handling.html
```

File: library/core/src/convert/mod.rs [hunk 8]
```diff
@@ -435,7 +417,7 @@ pub trait TryInto<T>: Sized {
 /// # Generic Implementations
 ///
 /// - `TryFrom<T> for U` implies [`TryInto`]`<U> for T`
-/// - [`try_from`] is reflexive, which means that `TryFrom<T> for T`
+/// - [`TryFrom::try_from`] is reflexive, which means that `TryFrom<T> for T`
```
Comment (jyn514) [inline library/core/src/convert/mod.rs:420]: nit: I think this looked nicer before.

File: library/core/src/convert/mod.rs [hunk 9]
```diff
@@ -484,11 +466,8 @@ pub trait TryInto<T>: Sized {
 /// assert!(try_successful_smaller_number.is_ok());
 /// ```
 ///
-/// [`try_from`]: trait.TryFrom.html#tymethod.try_from
-/// [`TryInto`]: trait.TryInto.html
-/// [`i32::MAX`]: ../../std/i32/constant.MAX.html
+/// [`i32::MAX`]: crate::i32::MAX
```
Comment (jyn514) [inline library/core/src/convert/mod.rs:469]: This links to the module constants, which are deprecated.
```suggestion
```
https://doc.rust-lang.org/nightly/std/i32/

File: library/core/src/convert/mod.rs [hunk 10]
```diff
@@ -691,10 +669,8 @@ impl AsRef<str> for str {
 /// the two `impl`s will start to overlap
 /// and therefore will be disallowed by the language’s trait coherence rules.
 ///
-/// [`Ok`]: ../result/enum.Result.html#variant.Ok
-/// [`Result`]: ../result/enum.Result.html
-/// [`TryFrom`]: trait.TryFrom.html
-/// [`Into`]: trait.Into.html
+/// [`Ok`]: super::result::Result::Ok
+/// [`Result`]: super::result::Result
```
Comment (jyn514) [inline library/core/src/convert/mod.rs:673]: These are in the prelude.
```suggestion
```

File: library/core/src/iter/mod.rs [hunk 11]
```diff
@@ -154,14 +152,11 @@
 //! produce an iterator. What gives?
 //!
 //! There's a trait in the standard library for converting something into an
-//! iterator: [`IntoIterator`]. This trait has one method, [`into_iter`],
+//! iterator: [`IntoIterator`]. This trait has one method, [`IntoIterator::into_iter`],
```
Comment (jyn514) [inline library/core/src/iter/mod.rs:155]: I think this was more clear before.

File: library/core/src/iter/mod.rs [hunk 12]
```diff
@@ -214,24 +209,20 @@
 //! often called 'iterator adapters', as they're a form of the 'adapter
 //! pattern'.
 //!
-//! Common iterator adapters include [`map`], [`take`], and [`filter`].
+//! Common iterator adapters include [`Iterator::map`], [`Iterator::take`], and [`Iterator::filter`].
```
Comment (jyn514) [inline library/core/src/iter/mod.rs:212]: I think this was more clear before.

File: library/core/src/iter/mod.rs [hunk 13]
```diff
@@ -248,8 +239,8 @@
 //! do nothing unless consumed
 //! ```
 //!
-//! The idiomatic way to write a [`map`] for its side effects is to use a
-//! `for` loop or call the [`for_each`] method:
+//! The idiomatic way to write a [`Iterator::map`] for its side effects is to use a
+//! `for` loop or call the [`Iterator::for_each`] method:
```
Comment (jyn514) [inline library/core/src/iter/mod.rs:243]: All of these were more clear before IMO.

**Review by jyn514 — COMMENTED** [review]
Let's avoid repeating `Iterator::` everywhere; it's pretty clear here that's what we're talking about and the Iterator docs are already pretty long.

File: library/core/src/convert/mod.rs (hunk 9 above)
Comment (nixphix) [reply library/core/src/convert/mod.rs:469]: removing it in my local was breaking the doc build

Comment (nixphix) [reply library/core/src/convert/mod.rs]: build [failed](https://github.com/rust-lang/rust/pull/75969/checks?check_run_id=1034624872#step:23:1630) after removing this 

```sh
413 | /// giving the [`i64`]'s value modulo [`i32::MAX`]) or by simply returning
    |                                        ^^^^^^^^^^ cannot be resolved, ignoring
```

**Comment by jyn514** [comment]
```
error: `[i32::MAX]` cannot be resolved, ignoring it.
   --> library/core/src/convert/mod.rs:413:40
    |
413 | /// giving the [`i64`]'s value modulo [`i32::MAX`]) or by simply returning
    |                                        ^^^^^^^^^^ cannot be resolved, ignoring
```
Hmm ... This looks like either https://github.com/rust-lang/rust/pull/74489 or https://github.com/rust-lang/rust/pull/75649. If you wait for the beta bump today they should both. E fixed.

**Comment by jyn514** [comment]
Beta was bumped today, try rebasing and see if it fixes the CI failure?

**Comment by nixphix** [comment]
@jyn514 tried rebasing twice still filing for the same reason, should I revert the change?

**Comment by jyn514** [comment]
> @jyn514 tried rebasing twice still filing for the same reason, should I revert the change?

I fixed the issue in https://github.com/rust-lang/rust/pull/76093. You should revert the change in the meantime, since it will break with `doc --stage 0` until the next release in 6 weeks.

**Comment by nixphix** [comment]
@jyn514 checks passed, I have set a reminder to fix this later

**Comment by jyn514** [comment]
@bors r-

Sorry - I missed a broken link. This will no longer resolve: https://github.com/rust-lang/rust/pull/75969/files#diff-d95315c7bb4e51591e307e2659bad42bL42

**Comment by nixphix** [comment]
ok let me add `Some(Item)`

**Pull request merged**